src/core/ts/background-scripts/background.js/telemetry/Telemetry.ts

```diff
@@ -5,78 +5,149 @@
 import Glean from "@mozilla/glean/webext";
 import { custom } from "./generated/pings";
 import { config } from "../../../config";
+import {
+  modelLoadTime,
+  translationTime,
+  wordsPerSecond,
+} from "./generated/performance";
+import { fromLang, toLang } from "./generated/metadata";
+import {
+  displayed,
+  changeLang,
+  closed,
+  neverTranslateSite,
+  translate,
+  neverTranslateLang,
+  notNow,
+} from "./generated/infobar";
 
+/**
+ * This class contains general telemetry initialization and helper code and synchronous telemetry-recording functions.
+ *
+ * Synchronous methods here is important, since it is the only way to guarantee that multiple Glean API calls are
+ * executed sequentially and not interleaved with other asynchronous Telemetry recording.
+ * For more information, see: https://github.com/mozilla-extensions/bergamot-browser-extension/pull/76#discussion_r602128568
+ */
 export class Telemetry {
-  private static _instance = null;
-  private _metricsToSubmit;
-
   constructor() {
     const appId = config.telemetryAppId;
     Glean.initialize(appId, true, {
       debug: { logPings: config.telemetryDebugMode },
     });
-    this._metricsToSubmit = 0;
     console.info(
       `Telemetry: initialization completed with application ID ${appId}.`,
     );
-    // Disabled telemetry until opt-out-preference is respected (https://github.com/mozilla-extensions/bergamot-browser-extension/issues/51)
-    this.setUploadEnabled(false);
-    console.info(`Note: Telemetry upload is disabled in this build`);
   }
 
-  /**
-   * Provides access to Telemetry singleton.
-   */
-  public static get global(): Telemetry {
-    if (Telemetry._instance === null) {
-      Telemetry._instance = new Telemetry();
-    }
-    return Telemetry._instance;
+  public onInfoBarDisplayed(tabId: number, from: string, to: string) {
+    fromLang.set(from);
+    toLang.set(to);
+    displayed.record();
+    this.submit();
   }
 
-  /**
-   * Enables uploading Glean metrics to telemetry server.
-   * @param val bool
-   */
-  public setUploadEnabled(val: boolean) {
-    Glean.setUploadEnabled(val);
+  public onSelectTranslateFrom(tabId: number, newFrom: string, to: string) {
+    fromLang.set(newFrom);
+    toLang.set(to);
+    changeLang.record();
```

Comment:
I think from the analysis perspective it would be more convenient to keep `lang_from`, 'lang_to` metadata written as a separate metric for all metrics including events. This allows us to set up filters in Looker or other dashboards for any analysis in the same way. Also, we'll have more metadata later, for example, `telemetry_client_id`.

Reply:
Wait, maybe I misunderstood this idea. If it is specifically for ChangeFrom ChangeTo events, then these extra keys will be very handy because we already occupy `from` and `to` slots in metadata and we need one extra slot `from` for changing events, which we can put into extra.